research/train.py

```diff
@@ -148,9 +156,24 @@ def train(cfg: DictConfig) -> None:
         _lf_one(
             log_engine_output,
             {
```

Comment:
Probably shouldn't be a dictionary since it only allows for 1 of each LOG.OP. Aren't there cases where you want 2 of the same LOG.OP but do slightly different things? Like Logging to two separate files

Reply:
That's the current behavior. The key is the operation, each element in the value is processed separately. So for the `LOG_OP.SAVE_IN_DATA_FILE` operation, each element in the list is saved to a separate file

Reply:
So there is no operation for saving metrics 1,2,3,4 into 1 file?

Reply:
I guess that could be added, but if you wanted to have 1,4 then 2,3 in separte files you would either need 2 LOG.OP of the same type or do what you are kind of doing for Visdom and send in to object that will configure that logging

Reply:
I think for readability is that the logging function is send a list of operations and and it corresponding arguments:
```python
[
    {op: {
        arguments:
        }
    },
    {op: ....
    }
]
```

Reply:
> I guess that could be added, but if you wanted to have 1,4 then 2,3 in separte files you would either need 2 LOG.OP of the same type or do what you are kind of doing for Visdom and send in to object that will configure that logging

Correct, currently there is no way to save multiple numbers to the same file. I'm not even sure what the file format people normally want. Binary to save space, CSV?

Not sure how readability improves. Other than jointly saving various outputs to combinations of files, what really improves. And if you're saving non-image outputs to files, why do you want more than 1 number per file? And in what format? I'm trying to understand use cases

Reply:
CSV is a pretty popular format for just data keeping since it is easy to use: throw into tables, throw into pandas, throw into numpy

Reply:
Also some people might want to save images that are attached to each other, like putting them side by side. So multiple of the same logging operations are needed or some more comprehensive configuration for the way images are logged. Some might want to log images plain, images with drawings, images  concat, or any other kind of operation done before saving them

Reply:
A couple of different things to do then.
- change file saving format to CSV
- change log operations from Dict to List

> CSV is a pretty popular format for just data keeping since it is easy to use: throw into tables, throw into pandas, throw into numpy

For the image manipulations you describe we're going to handle this in a few ways:
1. @mchiquier is writing some functions that can join images in helpful ways (based on torchvision, etc.)
2. Visdom supports logging multiple images in a grid
3. People can do whatever image manipulations they want before, the logging just wants an np.ndarray

Reply:
I've incorporated your thoughts on switching away from log operations as a dict. Creating a new issue to cover CSV logging.